src/network/local_socket_reply.cpp

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright (C) 2020 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "local_socket_reply.h"
+
+#include <multipass/format.h>
+
+namespace mp = multipass;
+
+namespace
+{
+constexpr int len = 65536;
+
+// The following is mostly "borrowed" from Qt src/network/access/qhttpthreaddelegate.cpp
+QNetworkReply::NetworkError statusCodeFromHttp(int httpStatusCode)
+{
+    QNetworkReply::NetworkError code;
+
+    // Only switch on the HTTP errors LXD will respond with for now
+    switch (httpStatusCode)
+    {
+    case 400: // Bad Request
+        code = QNetworkReply::ProtocolInvalidOperationError;
+        break;
+
+    case 401: // Authorization required
+        code = QNetworkReply::AuthenticationRequiredError;
+        break;
+
+    case 403: // Access denied
+        code = QNetworkReply::ContentAccessDenied;
+        break;
+
+    case 404: // Not Found
+        code = QNetworkReply::ContentNotFoundError;
+        break;
+
+    case 409: // Resource Conflict
+        code = QNetworkReply::ContentConflictError;
+        break;
+
+    case 500: // Internal Server Error
+        code = QNetworkReply::InternalServerError;
+        break;
+
+    default:
+        if (httpStatusCode > 500)
+        {
+            // some kind of server error
+            code = QNetworkReply::UnknownServerError;
+        }
+        else
+        {
+            // content error we did not handle above
+            code = QNetworkReply::UnknownContentError;
+        }
+    }
+
+    return code;
+}
+} // namespace
+
+mp::LocalSocketReply::LocalSocketReply(const QString& socket_path, const QNetworkRequest& request,
+                                       QIODevice* outgoingData)
+    : QNetworkReply(), local_socket{std::make_unique<QLocalSocket>(this)}, reply_data{QByteArray(len, '\0')}
```

Comment:
I ended up having to do this because the data we get from LXD is not terminated by `\0` and rather than trying to figure out where to add `\0` in the reply data, it was cleaner just setting the whole `QByteArray` to `\0` upon initialization.